{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam \n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import sys\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import argparse\n",
    "import glob\n",
    "\n",
    "import tensorflow as tf\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "from dagmm import DAGMM\n",
    "\n",
    "from scipy.special import logsumexp\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## set variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_bam = '/zfssz6/ST_MCHRI/BIGDATA/P18Z10200N0124/NIPT_CNV/final_file/NA12878/SRR098401_1.sorted.rmdup.realign.BQSR.bam'\n",
    "ref_fasta = '/zfssz6/ST_MCHRI/BIGDATA/database/BGI-seq500_OSS_download/human_reference/hg38/Homo_sapiens_assembly38.fasta'\n",
    "\n",
    "# varaible set\n",
    "bin_dir = '/zfssz6/ST_MCHRI/BIGDATA/P18Z10200N0124/NIPT_CNV/NIPT_CNV/GMM_1cnn_MODEL_train/bin'\n",
    "base_dir = '/zfssz6/ST_MCHRI/BIGDATA/P18Z10200N0124/NIPT_CNV/NIPT_CNV/GMM_1cnn_MODEL_train/run/winsize_1000k'\n",
    "cp_ratio = base_dir + '/test_sample/SRR098401_1.sorted.rmdup.realign.BQSR.bam.NA12878_winsize_1000000.correct_ratio.txt'\n",
    "model_dir = base_dir + '/model_dir'\n",
    "truth_cnv = bin_dir + '/ALL.wgs.mergedSV.v8.20130502.svs.genotypes.GRCh38.vcf.cnvs.dat'\n",
    "\n",
    "chrom_ref = ['chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10',\n",
    "'chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chr19','chr20','chr21','chr22','chrX','chrY','chrM']\n",
    "\n",
    "colors={0:'r', 1:'g', 2: 'b', 3:'k',4:'m',5:'c',6:'y'}\n",
    "legends={0:'DUP', 1: 'NEU', 2: 'DEL', 3:'DUP/DEL'}\n",
    "bin_size=[10000,100000,1000000]\n",
    "win_size=[10,15,20,30,50]\n",
    "win = bin_size[2]\n",
    "chrs = ['chr'+str(i) for i in range(1,23)] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## define function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rolling_window(a, window):\n",
    "    shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)\n",
    "    strides = a.strides + (a.strides[-1],)\n",
    "    return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_copy(cp_ratio, npysave):\n",
    "\n",
    "    ws = win_size[2]\n",
    "    cy_ratio_mat = []\n",
    "    cp_df = pd.read_csv(cp_ratio, sep='\\t')\n",
    "\n",
    "    for chrn in chrs:\n",
    "        chr_df = cp_df[(cp_df['space']==chrn) & (cp_df['copy'])].sort_values(by=['start'])\n",
    "        i_copy = chr_df['copy'].values\n",
    "        # add 0 at the beginning of sequence\n",
    "        tmp_copy = np.append([0]*(ws-1),i_copy)\n",
    "        i_cy_matrix= rolling_window(tmp_copy,ws)\n",
    "        if len(cy_ratio_mat) == 0:\n",
    "            cy_ratio_mat = i_cy_matrix\n",
    "        else:\n",
    "            cy_ratio_mat = np.concatenate((cy_ratio_mat,i_cy_matrix),axis=0)\n",
    "\n",
    "    np.save(os.path.join(npysave, cp_ratio.split(\"/\")[-1] + '_bins20.npy'), cy_ratio_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_model(model_dir):\n",
    "    \n",
    "    model_dagmm = DAGMM(\n",
    "        comp_hiddens=[32,16,1], comp_activation=tf.nn.tanh, # 16\n",
    "        est_hiddens=[8,16,3], est_activation=tf.nn.tanh, est_dropout_ratio=0.25,\n",
    "        epoch_size=100, minibatch_size=1024,normalize=False #1000\n",
    "    )\n",
    "    model_dagmm.restore(model_dir)\n",
    "    return model_dagmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_raw_copy(cp_ratio):\n",
    "\n",
    "    ws = win_size[2]\n",
    "    test_df = pd.read_csv(cp_ratio, sep='\\t')\n",
    "    \n",
    "    pred_chrms={}\n",
    "    for chrn in chrs:\n",
    "\n",
    "        chr_df = test_df[(test_df['space']==chrn) & (test_df['copy'])].sort_values(by=['start'])\n",
    "        i_copy = chr_df['copy'].values\n",
    "        start_pos=chr_df['start'].values\n",
    "        end_pos=chr_df['end'].values\n",
    "        reads = chr_df['reads'].values\n",
    "        gcs = chr_df['gc'].values\n",
    "        mapps = chr_df['map'].values\n",
    "    #     print(chrn)\n",
    "\n",
    "        d={'start':start_pos,'end':end_pos,'copy':i_copy,'gc':gcs,'reads':reads,'maps':mapps}\n",
    "        pred_chrms[chrn] = pd.DataFrame(data=d)\n",
    "    return pred_chrms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pred(cp_ratio, model_dagmm):\n",
    "    \n",
    "    ws = win_size[2]\n",
    "    test_df = pd.read_csv(cp_ratio, sep='\\t')\n",
    "     \n",
    "    pred_chrms={}\n",
    "    for chrn in chrs:\n",
    "\n",
    "        chr_df = test_df[(test_df['space']==chrn) & (test_df['copy'])].sort_values(by=['start'])\n",
    "        i_copy = chr_df['copy'].values\n",
    "        start_pos=chr_df['start'].values\n",
    "        end_pos=chr_df['end'].values\n",
    "        reads = chr_df['reads'].values\n",
    "        gcs = chr_df['gc'].values\n",
    "        mapps = chr_df['map'].values\n",
    "        # add 0 at the beginning of sequence\n",
    "        tmp_copy = np.append([0]*(ws-1),i_copy)\n",
    "        i_cy_matrix= rolling_window(tmp_copy,ws)   \n",
    "\n",
    "\n",
    "        energy,logit_probs = model_dagmm.predict(i_cy_matrix)\n",
    "        lse = logsumexp(logit_probs,axis=0)\n",
    "        log_resp=logit_probs-lse[np.newaxis,:]\n",
    "        comp_prob = np.exp(log_resp)\n",
    "        i_comp=np.argmax(comp_prob,axis=0)\n",
    "\n",
    "\n",
    "        d={'start':start_pos,'end':end_pos,'pred':i_comp,'copy':i_copy,'gc':gcs,'reads':reads,'maps':mapps}\n",
    "        pred_chrms[chrn] = pd.DataFrame(data=d)\n",
    "        \n",
    "    return pred_chrms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'HG01136'\n",
    "def get_turth_cnv(truth_cnv, sampleId):\n",
    "    \n",
    "    cnvs_annos = pd.read_csv(truth_cnv,\n",
    "               sep='\\t',usecols=['CHROM', 'POS', 'ID', 'ALT', 'SVTYPE',\n",
    "           'SVLEN', 'END','AF', 'SUPP_N_HET','SUPP_N_HOM_ALT','SUPP_HETS','SUPP_HOM_ALTS'])\n",
    "\n",
    "    test_cnvs = cnvs_annos[cnvs_annos['SUPP_HETS'].str.contains(sampleId) \n",
    "                           | cnvs_annos['SUPP_HOM_ALTS'].str.contains(sampleId)]\n",
    "\n",
    "    test_cnvs.sort_values(by=['CHROM','POS'],inplace=True)\n",
    "    test_cnvs.reset_index(drop=True, inplace=True)\n",
    "\n",
    "\n",
    "    adjstart = np.floor(test_cnvs['POS'].values*1.0/win) * win + 1\n",
    "    adjstart = adjstart.astype('int')\n",
    "    test_cnvs['ADJPOS']= adjstart\n",
    "    adjend = np.ceil(test_cnvs['END'].values*1.0/win) * win\n",
    "    adjend = adjend.astype('int')\n",
    "    test_cnvs['ADJEND'] = adjend\n",
    "    test_cnvs['NWIN']=(test_cnvs['ADJEND']-test_cnvs['ADJPOS']+1)//win \n",
    "    test_cnvs['ADJTYPE']=test_cnvs['SVTYPE']\n",
    "    test_cnvs.loc[(test_cnvs['SVTYPE']=='CNV') & ( ~test_cnvs['ALT'].str.contains('CN0')),'ADJTYPE']='DUP'\n",
    "    re_str = '('+sampleId+',[0-9]\\|[0-9])'\n",
    "    test_cnvs['TARGET_HET_GT']=test_cnvs['SUPP_HETS'].str.extract(re_str)\n",
    "    test_cnvs['TARGET_HOM_GT']=test_cnvs['SUPP_HOM_ALTS'].str.extract(re_str)\n",
    "    test_cnvs = test_cnvs[['CHROM','POS','ID','ALT','ADJTYPE','END','SVLEN','AF','ADJPOS','ADJEND','NWIN','TARGET_HET_GT','TARGET_HOM_GT']]\n",
    "    \n",
    "    return test_cnvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_cnv(test_cnvs,pred_chrms,chromosom):\n",
    "    \n",
    "    tru_chr1_cnvs = test_cnvs[test_cnvs['CHROM']==chromosom]\n",
    "    pred_chr1_cnv=pred_chrms['chr' + str(chromosom)]\n",
    "    pred_chr1_cnv['truth']='NEU'\n",
    "    pred_chr1_cnv['ID']=''\n",
    "    pred_chr1_cnv['svlen']=0\n",
    "    pred_chr1_cnv['POS']=0\n",
    "    pred_chr1_cnv['END']=0\n",
    "    pred_chr1_cnv['TARGET_HET_GT']=''\n",
    "    pred_chr1_cnv['TARGET_HOM_GT']=''\n",
    "\n",
    "\n",
    "    for i,x in tru_chr1_cnvs.iterrows():\n",
    "\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'truth'] = x['ADJTYPE']\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'ID'] = x['ID']\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'svlen'] = int(x['SVLEN'])\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'AF'] = x['AF']\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'POS'] = x['POS']\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'END'] = int(x['END'])\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'TARGET_HET_GT'] = x['TARGET_HET_GT']\n",
    "        pred_chr1_cnv.loc[(pred_chr1_cnv['start']>=x['ADJPOS']) & (pred_chr1_cnv['end']<=x['ADJEND']),'TARGET_HOM_GT'] = x['TARGET_HOM_GT']\n",
    "    \n",
    "    return pred_chr1_cnv[pred_chr1_cnv['truth']=='NEU']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def seq2num(seqs):\n",
    "    \n",
    "    tar_list = {'':0, 'N':0, 'A':1,'T':2,'C':3,'G':4}\n",
    "    results = []\n",
    "    \n",
    "    for base in seqs:\n",
    "        result = int(tar_list[base.upper()]) \n",
    "        results.append(result)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_GC(seqs):\n",
    "    \n",
    "    tar_list = {'':0, 'G':1, 'C':1,'A':-1,'T':-1,'N':0}\n",
    "    results = []\n",
    "    \n",
    "    for base in seqs:\n",
    "        result = int(tar_list[base.upper()]) \n",
    "        results.append(result)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate feature\n",
    "def gen_feature(chr, start, end, samfile, reffile):\n",
    "\n",
    "    bin_size = end - start\n",
    "\n",
    "    #variable initial\n",
    "    feature_list = ['feature_1','feature_2', 'feature_3', 'feature_3', 'feature_4', 'feature_5', 'feature_10', 'feature_11']\n",
    "    feature_1 = np.zeros(bin_size, int)\n",
    "    feature_2 = np.zeros(bin_size, int)\n",
    "    feature_3 = np.zeros(bin_size, int)\n",
    "    feature_4= np.zeros(bin_size, int)\n",
    "    feature_5= np.zeros(bin_size, int)\n",
    "\n",
    "    feature_10= np.zeros(bin_size, int)\n",
    "    feature_11= np.zeros(bin_size, int)\n",
    "\n",
    "    positions = np.array(range(start,end))\n",
    "\n",
    "    #check if no reads maped within regions\n",
    "    bin_readcounts = samfile.count(chr, start, end)\n",
    "    if bin_readcounts > 1:\n",
    "\n",
    "        #get ref message\n",
    "        ref_bin = reffile.fetch(chr, start, end)\n",
    "        base_ref_sequences = seq2num(list(ref_bin))\n",
    "        base_ref_is_gc = check_GC(list(ref_bin))                                \n",
    "\n",
    "        feature_5= np.array(base_ref_is_gc)\n",
    "        feature_10= np.array(base_ref_sequences)\n",
    "\n",
    "        #get sample message\n",
    "\n",
    "        iter = samfile.pileup(chr, start, end, truncate = True)\n",
    "        index = 0\n",
    "\n",
    "        for pileupcolumn in iter:\n",
    "\n",
    "            pos = pileupcolumn.pos\n",
    "            base_num_aligned = np.mean(pileupcolumn.get_num_aligned())\n",
    "            \n",
    "            if base_num_aligned == 0:\n",
    "                \n",
    "                #position without reads coverage\n",
    "                base_seq_qualities = 0\n",
    "                base_mapping_qualities = 0\n",
    "                base_query_sequences = 0\n",
    "                base_is_gc = 0\n",
    "                \n",
    "            else:\n",
    "                \n",
    "#                 print('query_sequences: ', pileupcolumn.get_query_sequences())\n",
    "#                 print(len(pileupcolumn.get_query_sequences()))\n",
    "#                 print('ok')\n",
    "\n",
    "                base_query_sequences = 0\n",
    "                base_is_gc = stats.mode(check_GC(pileupcolumn.get_query_sequences()))[0][0]\n",
    "\n",
    "\n",
    "                base_query_sequences = stats.mode(seq2num(pileupcolumn.get_query_sequences()))[0][0]\n",
    "                base_is_gc = stats.mode(check_GC(pileupcolumn.get_query_sequences()))[0][0]\n",
    "                    \n",
    "#                 print('query_sequences', pileupcolumn.get_query_sequences())\n",
    "#                 print('mapping_qualities: ',pileupcolumn.get_mapping_qualities())\n",
    "#                 print('query_qualities()',pileupcolumn.get_query_qualities())\n",
    "                base_seq_qualities =  np.mean(pileupcolumn.get_query_qualities())\n",
    "                base_mapping_qualities = np.mean(pileupcolumn.get_mapping_qualities())\n",
    "                \n",
    "                \n",
    "                \n",
    "\n",
    "            positions[index] = pos\n",
    "            feature_1[index] = base_num_aligned\n",
    "            feature_2[index] = base_seq_qualities\n",
    "            feature_3[index] = base_mapping_qualities\n",
    "            feature_4[index] = base_is_gc\n",
    "            feature_11[index] = base_query_sequences\n",
    "            index = index + 1\n",
    "\n",
    "        feature = dict(zip(feature_list, [sys._getframe().f_back.f_locals[a] for a in feature_list]))\n",
    "    else:\n",
    "\n",
    "        feature = 'Inf'\n",
    "\n",
    "    return feature\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3248: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/zfssz5/BC_PS/liuyu6/software/anaconda3/envs/ai/lib/python3.6/site-packages/ipykernel_launcher.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "test_cnvs = get_turth_cnv(truth_cnv, 'NA12878')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>11622816</td>\n",
       "      <td>11623132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>21459925</td>\n",
       "      <td>21460202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>35161503</td>\n",
       "      <td>35161830.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>36267681</td>\n",
       "      <td>36267914.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>58278238</td>\n",
       "      <td>58279154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>67542287</td>\n",
       "      <td>67543128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>72300660</td>\n",
       "      <td>72346132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>72312854</td>\n",
       "      <td>72318911.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>85514778</td>\n",
       "      <td>85540461.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>109644537</td>\n",
       "      <td>109648770.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS          END\n",
       "4      1   11622816   11623132.0\n",
       "7      1   21459925   21460202.0\n",
       "12     1   35161503   35161830.0\n",
       "13     1   36267681   36267914.0\n",
       "20     1   58278238   58279154.0\n",
       "22     1   67542287   67543128.0\n",
       "23     1   72300660   72346132.0\n",
       "24     1   72312854   72318911.0\n",
       "28     1   85514778   85540461.0\n",
       "37     1  109644537  109648770.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use_bin.head(10)\n",
    "chromosom = 1\n",
    "\n",
    "test_cnvs['AF'] = test_cnvs['AF'].str.split(',',expand=True)[0]\n",
    "AF_test_cnvs = test_cnvs[test_cnvs['AF'].astype(np.float) >= 0.5]\n",
    "use_bin = AF_test_cnvs[['CHROM', 'POS', 'END']]\n",
    "use_cnv_bin = use_bin[use_bin['CHROM']==chromosom]\n",
    "\n",
    "# print(len(use_bin))\n",
    "pred_chrms = get_raw_copy(cp_ratio)\n",
    "use_normal = get_normal_cnv(test_cnvs,pred_chrms,chromosom)\n",
    "use_normal_bin = use_normal[['start', 'end']]\n",
    "use_cnv_bin.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(use_cnv_bin))\n",
    "print(len(use_normal_bin))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read file\n",
    "samfile =  pysam.AlignmentFile(test_bam, 'rb')\n",
    "reffile =  pysam.FastaFile(ref_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generage cnv region features\n",
    "\n",
    "cnv_feature = {}\n",
    "cnv_reature_len = []\n",
    "for index, row in use_cnv_bin.iterrows():\n",
    "\n",
    "    contig = 'chr' + str(row['CHROM'])\n",
    "    start = row['POS']\n",
    "    end = int(row['END'])\n",
    "    \n",
    "    bin_position = contig + ':' + str(start) + '-' + str(end)\n",
    "\n",
    "    feature = gen_feature(contig, start, end, samfile, reffile)\n",
    "    cnv_reature_len.append(len(feature))\n",
    "    cnv_feature[bin_position] = feature\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generage normal region features\n",
    "normal_feature= {}\n",
    "normal_feature_len = []\n",
    "for index, row in use_normal_bin.iterrows():\n",
    "\n",
    "    contig = 'chr' + str(chromosom)\n",
    "    start = row['start']\n",
    "    end = int(row['end'])\n",
    "    \n",
    "    bin_position = contig + ':' + str(start) + '-' + str(end)\n",
    "\n",
    "    feature = gen_feature(contig, start, end, samfile, reffile)\n",
    "    normal_feature_len.append(len(feature))\n",
    "    normal_feature[bin_position] = feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(cnv_reature_len))\n",
    "print(len(normal_feature_len))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
